clap_complete/src/shells/zsh.rs

```diff
@@ -622,18 +622,49 @@ fn write_positionals_of(p: &Command) -> String {
 
     let mut ret = vec![];
 
-    for arg in p.get_positionals() {
+    // The index of the 2nd-to-last positional argument.
+    let positionals: Vec<_> = p.get_positionals().collect();
+    let penultimate_idx = if positionals.len() > 1 {
+        Some(positionals.len() - 2)
+    } else {
+        None
+    };
+    // Completions for commands that end with two Vec arguments require special care:
+    // - You can have two Vec args separated with a value_terminator.
+    // - You can have two Vec args with the second one set to last (raw sets last)
+    //   which will force a separator of -- before the second.
+    // zsh requires a terminator in both of these situations.
+    let needs_terminator = match positionals.last() {
+        Some(arg) => arg.is_last_set() || arg.get_num_args().expect("built").max_values() > 1,
```

Comment:
> max_values() > 1,

I don't see how this is relevant for if `--` is needed.

Reply:
After reading the zsh docs and re-reading the zsh completion code I think there's an even simpler solution.

We are currently using this option to `_arguments`:

```
-S: Do not complete anything after '--' and treat those as argument values
```

and remember that the core issue is that we are emitting two sets of `*::arguments` while zsh can only have one catch-all for `*::args` for positionals.

If we combine those two constraints, it *seems* like the solution is simple: special-case the *final* multi-valued argument (eg. the `raw` one) and simply don't emit it.

It could even be brain-dead and keep track of whether it's already seen a multi-valued argument without an explicit value_terminator and then simply ignore any of the remaining multi-valued arguments.

zsh already stops completing after the `--` due to our use of `-S`, so this seems like the simplest fix and I suspect the code will also end up being simpler.

I *think* that would work. I'll wait to hear back from you before I rebase and update the commit, though, in case there's something I'm missing.

Reply:
I leave it to people who who zsh and can verify the behavior.  If you can just simply say what the before and after behavior is for the end user, I'll sign off on it if it sounds reasonable.

Reply:
btw an alternative to fighting all of this is to help out on https://github.com/clap-rs/clap/issues/3166 which I hope to replace the shell-specific completions with.

Reply:
Thanks @epage I think the small tweaks in this PR should be simpler and more acceptable now.

For the end user the before behavior is that they would see an error when completing commands that have two multi-valued arguments. eg. when completing for a garden sub-command:

```
$ garden cmd -<TAB>
_arguments:comparguments:325: doubled rest argument definition: *::arguments -- Arguments to forward to custom commands:
```

After these changes the completion no longer generates errors and is able to offer completions for `garden cmd -<TAB>`.

Hopefully we're on the right path for now. #3166 looks like a great long-term solution, though.